src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,447 @@
+
+@[simp] lemma zero_def : zero = (0 : E⟮K⟯) := rfl
+
+end zero
+
+section negation
+
+variables {x y : K} (h : E.weierstrass_equation x y)
+
+include h
+
+/-- The $y$-coordinate of the negation of an affine point. -/
+@[nolint unused_arguments] def neg_y : K := -y - ↑E.a₁ * x - ↑E.a₃
+
+@[simp] lemma neg_y_def : neg_y h = -y - ↑E.a₁ * x - ↑E.a₃ := rfl
+
+/-- The negation of an affine point in `E` lies in `E`. -/
+lemma weierstrass_equation_neg : E.weierstrass_equation x $ neg_y h :=
+by { rw [weierstrass_equation_iff] at h, rw [weierstrass_equation_iff, neg_y_def, ← h], ring1 }
+
+omit h
+
+/-- Negation in `E⟮K⟯`. Given `P : E⟮K⟯`, use `-P` instead of `neg P`. -/
+def neg : E⟮K⟯ → E⟮K⟯
+| 0            := 0
+| (some x y h) := some x (neg_y h) $ weierstrass_equation_neg h
+
+instance : has_neg E⟮K⟯ := ⟨neg⟩
+
+@[simp] lemma neg_def (P : E⟮K⟯) : neg P = -P := rfl
+
+@[simp] lemma neg_zero : -0 = (0 : E⟮K⟯) := rfl
+
+@[simp] lemma neg_some : -some x y h = some x (neg_y h) (weierstrass_equation_neg h) := rfl
+
+@[simp] lemma neg_neg (P : E⟮K⟯) : -(-P) = P := by { cases P, { refl }, { simp, ring1 } }
+
+instance : has_involutive_neg E⟮K⟯ := ⟨neg, neg_neg⟩
+
+end negation
+
+end point
+
+end basic
+
+open_locale EllipticCurve
+
+variables {F : Type u} [field F] {E : EllipticCurve F} {K : Type v} [field K] [algebra F K]
+
+section dbl_add
+
+/-!
+### The addition law of `E⟮K⟯`
+
+Given `P Q : E⟮K⟯`, use `P + Q` instead of `add P Q`.
+-/
+
+namespace point
+
+section doubling
+
+variables {x y : K} (h : E.weierstrass_equation x y)
+
+include h
+
+/-- The gradient of the tangent line of `E` at an affine point $(x, y)$. This is not well-defined
+only in the case of $y = -y - a_1x - a_3$, which is precisely when the tangent is vertical. -/
+def dbl_L : K := (3 * x ^ 2 + 2 * ↑E.a₂ * x + ↑E.a₄ - ↑E.a₁ * y) / (y - neg_y h)
+
+@[simp] lemma dbl_L_def :
+  dbl_L h = (3 * x ^ 2 + 2 * ↑E.a₂ * x + ↑E.a₄ - ↑E.a₁ * y) / (y - neg_y h) :=
+rfl
+
+/-- The $x$-coordinate of the doubling of an affine point whose tangent is not vertical. -/
+def dbl_x : K := dbl_L h ^ 2 + ↑E.a₁ * dbl_L h - ↑E.a₂ - 2 * x
+
+@[simp] lemma dbl_x_def : dbl_x h = dbl_L h ^ 2 + ↑E.a₁ * dbl_L h - ↑E.a₂ - 2 * x := rfl
+
+/-- The $y$-coordinate of the doubling of an affine point whose tangent is not vertical,
+before applying the final negation that maps $y$ to $-y - a_1x - a_3$. -/
+def dbl_y' : K := dbl_L h * (dbl_x h - x) + y
+
+@[simp] lemma dbl_y'_def : dbl_y' h = dbl_L h * (dbl_x h - x) + y := rfl
+
+/-- The $y$-coordinate of the doubling of an affine point whose tangent is not vertical. -/
+def dbl_y : K := -dbl_y' h - ↑E.a₁ * dbl_x h - ↑E.a₃
+
+@[simp] lemma dbl_y_def : dbl_y h = -dbl_y' h - ↑E.a₁ * dbl_x h - ↑E.a₃ := rfl
+
+/-- The doubling of an affine point in `E` whose tangent is not vertical,
+before applying the final negation that maps $y$ to $-y - a_1x - a_3$, lies in `E`. -/
+lemma weierstrass_equation_dbl' (hy : y ≠ neg_y h) : E.weierstrass_equation (dbl_x h) (dbl_y' h) :=
```

Comment:
I think we should use a more conceptual proof to unify this and `weierstrass_equation_add'` below. In both case we plug in `Y := L(X-x₁)+y₁` into the Weierstrass polynomial `F(X,Y)` to get `f(X)`, take the negative of the `X^2` coefficient, and subtract two roots from it to get another root. Clearly, `x` is a root of `f(X)` iff `(x, L(x-x₁)+y₁)` lies on the elliptic curve. When `x₁ ≠ x₂` then `f(X)` certainly has the factor `(X-x₁)(X-x₂)`, but when `x = x₁ = x₂` and `y = y₁ = y₂` we need that `f'(x) = 0` in order for `f(X)` to have the factor `(X-x)^2`. For this we need a multivariate chain rule: `f'(x)` is the derivative of `F(X, L(X-x)+y)` at `X=x`, which is equal to the partial derivative `F_X(X,Y)` evaluated at `(x, L(x-x)+y) = (x, y)` plus `L` times `F_Y(X,Y)` evaluated at the same point, which is zero by definition of `L` (`dbl_L`).

Reply:
I thought for a while and I question if it is worth doing all these (reinterpreting the goal, defining the partial derivative, proving the multivariate chain rule) just to unify these two proofs. I just realised (after proving [the nonsingularity condition](https://github.com/leanprover-community/mathlib/blob/ebe7022d68057adb2efb360f8a682faa9ff56798/src/algebraic_geometry/elliptic_curve/weierstrass.lean#L331)) that we can also do something similar here, where you prove the result for (0, 0) and apply a change of variables to get the same result for (x, y), and it involves no large expressions at all! Would you be OK with this?